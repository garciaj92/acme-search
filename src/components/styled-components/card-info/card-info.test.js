import React from "react";
import CardInfo from "../card-info";

import { mount, shallow } from "enzyme";
import toJson from 'enzyme-to-json';

// Mocks
import { consoleErrorFn } from "../../../__mocks__/console.mock";

const props = {
    iconSrc: "./test-icon.svg",
    heading: "Test Heading",
    href: "https://github.com/garciaj92",
    content: "Test Content"
}

describe("CardInfo: Test visual components", () => {
    let wrapper;
    beforeEach(() => {
        wrapper = mount(<CardInfo {...props} />);
    });

    it("renders correctly", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("displays sets the props correctly", () => {
        expect(wrapper.find("img").first().prop("src")).toEqual(props.iconSrc);
        expect(wrapper.find("h5").first().text()).toEqual(props.heading);
        expect(wrapper.find("a").first().prop("href")).toEqual(props.href);
        expect(wrapper.find("a").first().text()).toEqual(props.content);
    });
});

describe("CardInfo: Test prop type validation", () => {
    let consoleErrorSpy;
    beforeEach(() => {
        console.error = consoleErrorFn
        consoleErrorSpy = jest.spyOn(console, "error");
    });

    afterEach(() => {
        consoleErrorSpy.mockClear();
    });
    
    it("does not throw an error when props are provided", () => {
        shallow(<CardInfo {...props} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    it("does not throw an error when href prop is not provided", () => {
        const { 
            iconSrc,
            heading,
            content
        } = props;
        shallow(<CardInfo iconSrc={iconSrc} heading={heading} content={content} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    // What is stated below applies to the remaining test cases in this describe block:
    //      console.error is called twice for each prop validator that fails because
    //      there are two components that use each failing prop. This means that if
    //      two props fail we have four errors.

    it("throws an error when iconSrc prop is not provided", () => {
        const { 
            heading,
            content
        } = props;
        shallow(<CardInfo heading={heading} content={content} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when heading prop is not provided", () => {
        const { 
            iconSrc,
            content
        } = props;
        shallow(<CardInfo iconSrc={iconSrc} content={content} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when content prop is not provided", () => {
        const { 
            iconSrc,
            heading,
        } = props;
        shallow(<CardInfo iconSrc={iconSrc} heading={heading} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error each prop that is not a `string`", () => {
        shallow(<CardInfo iconSrc={0} heading={0} content={0} href={0} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(8);
    });
});