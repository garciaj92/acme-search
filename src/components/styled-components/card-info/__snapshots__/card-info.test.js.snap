// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardInfo: Test visual components renders correctly 1`] = `
<CardInfo
  content="Test Content"
  heading="Test Heading"
  href="https://github.com/garciaj92"
  iconSrc="./test-icon.svg"
>
  <CardIconSubHeading
    heading="Test Heading"
    iconSrc="./test-icon.svg"
  >
    <div
      className="d-flex justify-content-center align-items-center my-2"
    >
      <Image
        fluid={false}
        rounded={false}
        roundedCircle={false}
        src="./test-icon.svg"
        thumbnail={false}
        width={30}
      >
        <img
          className=""
          src="./test-icon.svg"
          width={30}
        />
      </Image>
      <h5
        className="mb-0 mx-1"
      >
        Test Heading
      </h5>
    </div>
  </CardIconSubHeading>
  <CardLink
    content="Test Content"
    href="https://github.com/garciaj92"
  >
    <a
      className="text-muted mb-0 d-block"
      href="https://github.com/garciaj92"
      onClick={[Function]}
    >
      Test Content
    </a>
  </CardLink>
</CardInfo>
`;
